### PR TITLE
[master  <  T0740-MG] Add procedure for setting a kafka stream offset 

### PR DESCRIPTION
**Pull request by kostasrim**
[master < Epic] PR
- [ ] Check, and update documentation if necessary
- [ ] Update [changelog](https://docs.memgraph.com/memgraph/changelog)
- [ ] Write E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Update [changelog](https://docs.memgraph.com/memgraph/changelog)

### CODE REVIEW

File: src/integrations/kafka/consumer.cpp [hunk 1]
```diff
@@ -339,7 +347,21 @@ void Consumer::StartConsuming() {
...
         consumer_function_(batch);
-        if (const auto err = consumer_->commitSync(); err != RdKafka::ERR_NO_ERROR) {
+        std::vector<RdKafka::TopicPartition *> partitions;
+        utils::OnScopeExit clear_partitions([&p = partitions]() { RdKafka::TopicPartition::destroy(p); });
+
+        if (const auto err = consumer_->assignment(partitions); err != RdKafka::ERR_NO_ERROR) {
+          spdlog::warn("Saving the commited offset of consumer {} failed: {}", info_.consumer_name,
+                       RdKafka::err2str(err));
+          throw ConsumerCheckFailedException(
+              info_.consumer_name, fmt::format("Couldn't save commited offsets: '{}'", RdKafka::err2str(err)));
```
Comment (antaljanosbenjamin) [inline src/integrations/kafka/consumer.cpp:357]: This message is a bit off. "Committed offset" practically the saved offset, so I would rephrase this:
```suggestion
              info_.consumer_name, fmt::format("Couldn't get assignment to commit offsets: '{}'", RdKafka::err2str(err)));
```

File: src/integrations/kafka/consumer.hpp [hunk 2]
```diff
@@ -137,6 +140,13 @@ class Consumer final : public RdKafka::EventCb {
   /// Returns true if the consumer is actively consuming messages.
   bool IsRunning() const;
 
+  /// Sets the consumer's offset.
+  ///
+  /// This function returns the empty string on success or an error message otherwise.
+  ///
+  /// @param offset: the offset to set.
+  std::string SetConsumerOffsets(int64_t offset);
```
Comment (antaljanosbenjamin) [inline src/integrations/kafka/consumer.hpp:148]: I would change this to `[[nodiscard]] utils::BasicResult<std::string>`. It is more clear even without documentation and enforces the error check.

File: src/query/procedure/module.cpp [hunk 3]
```diff
@@ -325,15 +325,17 @@ bool SharedLibraryModule::Load(const std::filesystem::path &file_path) {
   dlerror();  // Clear any existing error.
   handle_ = dlopen(file_path.c_str(), RTLD_NOW | RTLD_LOCAL);
   if (!handle_) {
-    // NOLINTNEXTLINE(concurrency-mt-unsafe)
-    spdlog::error(utils::MessageWithLink("Unable to load module {}; {}.", file_path, dlerror(), "https://memgr.ph/modules"));
+    spdlog::error(utils::MessageWithLink("Unable to load module {}; {}.", file_path, dlerror(),
+                                         "https://memgr.ph/modules"));  // NOLINT(concurrency-mt-unsafe)
```
Comment (antaljanosbenjamin) [inline src/query/procedure/module.cpp:329]:
```suggestion
    spdlog::error(utils::MessageWithLink("Unable to load module {}; {}.", file_path, dlerror(),
                                         "https://memgr.ph/modules"));
```

I think the other PR has to be merged to master before this PR can be merged, because that disables this check.

File: src/query/procedure/module.cpp [hunk 4]
```diff
@@ -384,8 +386,8 @@ bool SharedLibraryModule::Close() {
     spdlog::warn("When closing module {}; mgp_shutdown_module returned {}", file_path_, shutdown_res);
   }
   if (dlclose(handle_) != 0) {
-    // NOLINTNEXTLINE(concurrency-mt-unsafe)
-    spdlog::error(utils::MessageWithLink("Failed to close module {}; {}.", file_path_, dlerror(), "https://memgr.ph/modules"));
+    spdlog::error(utils::MessageWithLink("Failed to close module {}; {}.", file_path_, dlerror(),
+                                         "https://memgr.ph/modules"));  // NOLINTN(concurrency-mt-unsafe)
```
Comment (antaljanosbenjamin) [inline src/query/procedure/module.cpp:390]:
```suggestion
    spdlog::error(utils::MessageWithLink("Failed to close module {}; {}.", file_path_, dlerror(),
                                         "https://memgr.ph/modules"));
```

File: src/query/procedure/module.hpp [hunk 5]
```diff
@@ -117,6 +117,8 @@ class ModuleRegistry final {
   /// Returns the shared memory allocator used by modules
   utils::MemoryResource &GetSharedMemoryResource() noexcept;
 
+  bool RegisterProcedure(const std::string_view name, mgp_proc proc);
```
Comment (antaljanosbenjamin) [inline src/query/procedure/module.hpp:120]: I would rename this to `RegisterBuiltinProcedure`, because this function cannot be used to register arbitrary procedures, only the "builtin" ones to the `mg` module.

File: src/query/streams.cpp [hunk 6]
```diff
@@ -185,7 +186,28 @@ Streams::Streams(InterpreterContext *interpreter_context, std::string bootstrap_
...
+    const auto *stream_name = procedure::Call<const char *>(mgp_value_get_string, arg_stream_name);
+    auto *arg_offset = procedure::Call<mgp_value *>(mgp_list_at, args, 1);
+    const auto offset = procedure::Call<int64_t>(mgp_value_get_int, arg_offset);
+    std::string error = ictx->streams.SetStreamOffset(stream_name, offset);
+    if (!error.empty()) {
+      MG_ASSERT(mgp_result_set_error_msg(result, error.c_str()) == MGP_ERROR_NO_ERROR);
+    }
+  };
+
+  const std::string proc_name = "set_stream_offset";
```
[the remark this answers is not in the document]
Comment (antaljanosbenjamin) [reply src/query/streams.cpp:202]: As we discussed dots are only possible in module names and not in procedure names. As a result we agreed to use `kafka_set_stream_offset`.

File: src/query/streams.cpp [hunk 7]
```diff
@@ -185,7 +186,28 @@ Streams::Streams(InterpreterContext *interpreter_context, std::string bootstrap_
...
+  mgp_proc proc("set_stream_offset", set_stream_offset, utils::NewDeleteResource(), false);
+  MG_ASSERT(mgp_proc_add_arg(&proc, proc_name.c_str(), procedure::Call<mgp_type *>(mgp_type_string)) ==
+            MGP_ERROR_NO_ERROR);
+  MG_ASSERT(mgp_proc_add_arg(&proc, proc_name.c_str(), procedure::Call<mgp_type *>(mgp_type_int)) ==
+            MGP_ERROR_NO_ERROR);
```
Comment (antaljanosbenjamin) [inline src/query/streams.cpp:207]:
```suggestion
  MG_ASSERT(mgp_proc_add_arg(&proc, "stream_name", procedure::Call<mgp_type *>(mgp_type_string)) ==
            MGP_ERROR_NO_ERROR);
  MG_ASSERT(mgp_proc_add_arg(&proc, "offset", procedure::Call<mgp_type *>(mgp_type_int)) ==
            MGP_ERROR_NO_ERROR);
```

File: src/query/streams.cpp [hunk 8]
```diff
@@ -185,7 +186,28 @@ Streams::Streams(InterpreterContext *interpreter_context, std::string bootstrap_
...
       bootstrap_servers_(std::move(bootstrap_servers)),
-      storage_(std::move(directory)) {}
+      storage_(std::move(directory)) {
+  auto set_stream_offset = [ictx = interpreter_context](mgp_list *args, mgp_graph * /*graph*/, mgp_result *result,
+                                                        mgp_memory * /*memory*/) {
+    auto *arg_stream_name = procedure::Call<mgp_value *>(mgp_list_at, args, 0);
+    const auto *stream_name = procedure::Call<const char *>(mgp_value_get_string, arg_stream_name);
+    auto *arg_offset = procedure::Call<mgp_value *>(mgp_list_at, args, 1);
+    const auto offset = procedure::Call<int64_t>(mgp_value_get_int, arg_offset);
+    std::string error = ictx->streams.SetStreamOffset(stream_name, offset);
```
Comment (antaljanosbenjamin) [inline src/query/streams.cpp:196]:
```suggestion
    const auto error = ictx->streams.SetStreamOffset(stream_name, offset);
```

File: src/query/streams.cpp [hunk 9]
```diff
@@ -185,7 +186,28 @@ Streams::Streams(InterpreterContext *interpreter_context, std::string bootstrap_
...
+    if (!error.empty()) {
+      MG_ASSERT(mgp_result_set_error_msg(result, error.c_str()) == MGP_ERROR_NO_ERROR);
```
Comment (antaljanosbenjamin) [inline src/query/streams.cpp:198]: I think setting the error message is just best effort thing and we should crash memgraph in case it doesn't succeed.

File: src/query/streams.cpp [hunk 10]
```diff
@@ -434,4 +456,14 @@ void Streams::Persist(StreamStatus &&status) {
...
 
 std::string_view Streams::BootstrapServers() const { return bootstrap_servers_; }
+
+std::string Streams::SetStreamOffset(const std::string_view stream_name, int64_t offset) {
+  auto lock_ptr = streams_.Lock();
+  if (auto it = lock_ptr->find(std::string(stream_name)); it != lock_ptr->end()) {
+    auto consumer_lock_ptr = it->second.consumer->Lock();
+    return consumer_lock_ptr->SetConsumerOffsets(offset);
+  }
+  return fmt::format("Stream: {} not found", stream_name);
```
Comment (antaljanosbenjamin) [inline src/query/streams.cpp:466]:
```suggestion
  auto it = GetStream(*locked_streams, stream_name);
  auto consumer_lock_ptr = it->second.consumer->Lock();
  return consumer_lock_ptr->SetConsumerOffsets(offset);
```

I don't like `std::string` return value for error handling in general, but here there is an extra argument against it: every other function throws an exception if the stream cannot be found. This function breaks the consistency.

File: tests/e2e/streams/streams_tests.py [hunk 11]
```diff
@@ -342,85 +367,137 @@ def test_check_already_started_stream(topics, connection):
...
+    )
+
+    messages = [f"{i} message" for i in range(1, 21)]
+
+    for message in messages:
+        producer.send(topics[0], message.encode()).get(timeout=60)
+
+    def execute_set_offset_and_consume(id):
+        common.execute_and_fetch_all(
+            cursor, f"CALL mg.set_stream_offset('test', {id}) YIELD result"
```
Comment (antaljanosbenjamin) [inline tests/e2e/streams/streams_tests.py:473]: What does `YIELD result` means here? As far as I can see, `set_stream_offset` doesn't have a return value.

File: tests/e2e/streams/streams_tests.py [hunk 12]
```diff
@@ -342,85 +367,137 @@ def test_check_already_started_stream(topics, connection):
...
+        )
+        common.start_stream(cursor, "test")
+
+        time.sleep(2)
```
Comment (antaljanosbenjamin) [inline tests/e2e/streams/streams_tests.py:477]: Maybe you can use `check_vertex_exists_with_topic_and_payload` from `common.py`:
`assert check_vertex_exists_with_topic_and_payload(cursor, topics[0], messages[-1])`

It will wait up to 10 seconds and polling the DB for every 0.1 second.

Comment (antaljanosbenjamin) [reply tests/e2e/streams/streams_tests.py:477]: I see it won't work when the offset is `-2`. Maybe try to use `check_one_result_row` or `timed_wait` from `common.py`, both of them should suit this scenario with a proper query.

File: tests/e2e/streams/streams_tests.py [hunk 13]
```diff
@@ -342,85 +367,137 @@ def test_check_already_started_stream(topics, connection):
...
+    common.execute_and_fetch_all(cursor, "MATCH (n) DETACH DELETE n")
+
+    res = execute_set_offset_and_consume(-1)
+    assert len(res) == 20
+
+    assert all([comparison_check(a, b) for a, b in zip(messages, res)])
+    res = common.execute_and_fetch_all(cursor, "MATCH (n) DETACH DELETE n")
+
+    res = execute_set_offset_and_consume(-2)
+    assert len(res) == 0
```
Comment (antaljanosbenjamin) [inline tests/e2e/streams/streams_tests.py:500]: Consider sending one more message to verify the stream can consume message after setting the offset to the latest.

Comment (kostasrim) [reply tests/e2e/streams/streams_tests.py:547]: Good idea, done!

File: tests/e2e/streams/streams_tests.py (hunk 12 above)
Comment (kostasrim) [reply tests/e2e/streams/streams_tests.py:513]: I am unsure how `timed_wait` here would help and I don't see how `check_one_result_row` makes this any simpler?

File: tests/e2e/streams/streams_tests.py (hunk 11 above)
Comment (kostasrim) [reply tests/e2e/streams/streams_tests.py]: that was mistyped and removed

File: src/query/procedure/module.hpp (hunk 5 above)
Comment (kostasrim) [reply src/query/procedure/module.hpp]: Good suggestion!

File: src/query/streams.cpp (hunk 9 above)
Comment (antaljanosbenjamin) [reply src/query/streams.cpp]: Yes, I wanted to say "shouldn't", but my brain was faster than my hands :smile:

File: tests/e2e/streams/streams_tests.py (hunk 12 above)
Comment (antaljanosbenjamin) [reply tests/e2e/streams/streams_tests.py:513]: I think `assert check_one_result_row(cursor, f"MATCH (n: MESSAGE {{payload: {messages[-1]}}}) RETURN n")` would do the trick: it might finish faster than 2 seconds if the stream is fast, but would wait up to 10 seconds. In is more robust in case of random slowdowns etc.

File: src/integrations/kafka/consumer.cpp [hunk 14]
```diff
@@ -339,7 +345,22 @@ void Consumer::StartConsuming() {
 
       try {
         consumer_function_(batch);
-        if (const auto err = consumer_->commitSync(); err != RdKafka::ERR_NO_ERROR) {
+        std::vector<RdKafka::TopicPartition *> partitions;
+        utils::OnScopeExit clear_partitions([&p = partitions]() { RdKafka::TopicPartition::destroy(p); });
```
Comment (antonio2368) [inline src/integrations/kafka/consumer.cpp:349]:
```suggestion
        utils::OnScopeExit clear_partitions([&]() { RdKafka::TopicPartition::destroy(p); });
```
Why not just this?

File: src/integrations/kafka/consumer.cpp [hunk 15]
```diff
@@ -339,7 +345,22 @@ void Consumer::StartConsuming() {
...
+        if (const auto err = consumer_->assignment(partitions); err != RdKafka::ERR_NO_ERROR) {
+          spdlog::warn("Saving the commited offset of consumer {} failed: {}", info_.consumer_name,
+                       RdKafka::err2str(err));
+          throw ConsumerCheckFailedException(
+              info_.consumer_name,
+              fmt::format("Couldn't get assignment to commit offsets: '{}'", RdKafka::err2str(err)));
+        }
+
+        for (auto *partition : partitions) {
+          partition->set_offset(batch.back().Offset() + 1);
```
Comment (antonio2368) [inline src/integrations/kafka/consumer.cpp:360]: Why don't we store the `batch.back().Offset()` into a variable?

File: src/query/procedure/module.cpp [hunk 16]
```diff
@@ -638,6 +645,16 @@ void ModuleRegistry::UnloadAllModules() {
...
 utils::MemoryResource &ModuleRegistry::GetSharedMemoryResource() noexcept { return *shared_; }
 
+bool ModuleRegistry::RegisterMgProcedure(const std::string_view name, mgp_proc proc) {
+  std::unique_lock<utils::RWLock> guard(lock_);
+  if (auto module = modules_.find("mg"); module != modules_.end()) {
+    auto *builtin_module = dynamic_cast<BuiltinModule *>(module->second.get());
+    builtin_module->AddProcedure(name, std::move(proc));
+    return true;
+  }
+  return false;
```
Comment (antonio2368) [inline src/query/procedure/module.cpp:655]:
```suggestion
  auto module = modules_.find("mg");
  if (module == modules_.end()) {
    return false;
  }
  auto *builtin_module = dynamic_cast<BuiltinModule *>(module->second.get());
  builtin_module->AddProcedure(name, std::move(proc));
  return true;
```
Maybe like this? :smile:

File: src/query/procedure/module.hpp [hunk 17]
```diff
@@ -117,6 +117,8 @@ class ModuleRegistry final {
   /// Returns the shared memory allocator used by modules
   utils::MemoryResource &GetSharedMemoryResource() noexcept;
 
+  bool RegisterMgProcedure(const std::string_view name, mgp_proc proc);
```
Comment (antonio2368) [inline src/query/procedure/module.hpp:120]:
```suggestion
  bool RegisterMgProcedure(std::string_view name, mgp_proc proc);
```
Clang-tidy didn't trigger on this?

File: tests/e2e/streams/streams_tests.py [hunk 18]
```diff
@@ -408,19 +472,84 @@ def test_bootstrap_server(producer, topics, connection, transformation):
...
+            cursor, "CALL mg.kafka_set_stream_offset('foo', 10)"
+        )
+
+    res = execute_set_offset_and_consume(10)
+    assert len(res) == 10
+
+    def comparison_check(a, b):
+        return a == str(b).strip("'(,)")
+
+    assert all([comparison_check(a, b) for a, b in zip(messages[10:], res)])
```
Comment (antonio2368) [inline tests/e2e/streams/streams_tests.py:537]:
```suggestion
    assert all(comparison_check(a, b) for a, b in zip(messages[10:], res))
```
Should work I think.

**Review by antonio2368 — CHANGES_REQUESTED** [review]
Nothing major, in general, it looks really good!

File: src/query/streams.cpp (hunk 9 above)
Comment (antaljanosbenjamin) [reply src/query/streams.cpp]: So here I propose to remove the `MG_ASSERT`.

File: src/integrations/kafka/consumer.cpp (hunk 15 above)
Comment (kostasrim) [reply src/integrations/kafka/consumer.cpp:360]: Because it's not used anywhere else?

File: src/query/procedure/module.cpp (hunk 16 above)
Comment (kostasrim) [reply src/query/procedure/module.cpp:654]: I am not sure about this. I personally like that the variable is local to the scope on my version but I am happy to make  the change if you find this snippet more readable.

File: src/query/procedure/module.hpp (hunk 17 above)
Comment (kostasrim) [reply src/query/procedure/module.hpp]: Don't know but good catch!

File: tests/e2e/streams/streams_tests.py (hunk 18 above)
Comment (kostasrim) [reply tests/e2e/streams/streams_tests.py]: I changed this code based on Benjo's suggestion

File: src/query/procedure/module.cpp (hunk 16 above)
Comment (antonio2368) [reply src/query/procedure/module.cpp:654]: It's fine, wanted to hear your opinion.

File: src/integrations/kafka/consumer.cpp (hunk 15 above)
Comment (antonio2368) [reply src/integrations/kafka/consumer.cpp:360]: I was going more for getting the same value from the back of the vector for each partition instead of getting it once and storing it into a variable.

File: tests/e2e/streams/streams_tests.py [hunk 19]
```diff
@@ -143,23 +157,27 @@ def test_check_stream(producer, topics, connection, transformation):
         producer.send(topics[0], message).get(timeout=60)
 
     def check_check_stream(batch_limit):
-        assert transformation == "transform.simple" \
+        assert (
+            transformation == "transform.simple"
             or transformation == "transform.with_parameters"
+        )
```
Comment (jbajic) [inline tests/e2e/streams/streams_tests.py:163]:
```suggestion
        assert transformation in TRANSFORMATIONS_TO_CHECK
```

File: tests/e2e/streams/streams_tests.py [hunk 20]
```diff
@@ -342,64 +395,75 @@ def test_check_already_started_stream(topics, connection):
...
+        lambda: common.get_is_running(cursor, "test_stream")
+    )
     common.start_stream(cursor, "test_stream")
     end = time.time()
 
-    assert (end - start) < 1.3 * \
-        timeout_ms, "The START STREAM was blocked too long"
+    assert (
+        end - start
+    ) < 1.3 * timeout_ms, "The START STREAM was blocked too long"
```
Comment (jbajic) [inline tests/e2e/streams/streams_tests.py:425]: Is 1.3 a magic number?

**Review by jbajic — APPROVED** [review]
Good job, minor comments :tada:

File: src/query/streams.cpp [hunk 21]
```diff
@@ -185,7 +186,27 @@ Streams::Streams(InterpreterContext *interpreter_context, std::string bootstrap_
...
+  auto set_stream_offset = [ictx = interpreter_context](mgp_list *args, mgp_graph * /*graph*/, mgp_result *result,
+                                                        mgp_memory * /*memory*/) {
+    auto *arg_stream_name = procedure::Call<mgp_value *>(mgp_list_at, args, 0);
+    const auto *stream_name = procedure::Call<const char *>(mgp_value_get_string, arg_stream_name);
+    auto *arg_offset = procedure::Call<mgp_value *>(mgp_list_at, args, 1);
+    const auto offset = procedure::Call<int64_t>(mgp_value_get_int, arg_offset);
+    const auto error = ictx->streams.SetStreamOffset(stream_name, offset);
+    if (error.HasError()) {
+      MG_ASSERT(mgp_result_set_error_msg(result, error.GetError().c_str()) == MGP_ERROR_NO_ERROR,
+                "kafka_set_offset() unable to set stream offset");
```
Comment (antaljanosbenjamin) [inline src/query/streams.cpp:199]:
```suggestion
                "kafka_set_stream_offset() unable to set stream offset");
```

File: tests/e2e/streams/streams_tests.py (hunk 19 above)
Comment (antaljanosbenjamin) [reply tests/e2e/streams/streams_tests.py:163]: I think this is my bad... I think I put this assert here to make the if statement below simpler. But I think this was a big mistake from my side, because it doesn't express the intent well enough. If you want to improve this, then I would remove this assert completely and change the if statement below to:
```python
    if transformation == "transformation.simple":
        # ....
    elif transformation == "transformation.with_parameters":
        # ....
    else:
        assert False
```
This way the intention is much clear.

File: tests/e2e/streams/streams_tests.py [hunk 22]
```diff
@@ -408,19 +472,78 @@ def test_bootstrap_server(producer, topics, connection, transformation):
...
+
+    def consume(expected_messages):
+        common.start_stream(cursor, "test")
+        if len(expected_messages) == 0:
+            time.sleep(2)
+
+        for message in expected_messages:
+            assert common.check_one_result_row(
+                cursor, f"MATCH (n: MESSAGE {{payload: '{message}'}}) RETURN n"
+            )
```
Comment (antaljanosbenjamin) [inline tests/e2e/streams/streams_tests.py:518]: I didn't communicated my idea clear enough. I would do something like this:
```suggestion
        if len(expected_messages) == 0:
            time.sleep(2)
        else:
            assert check_one_result_row(cursor, f"MATCH (n: MESSAGE {{payload: {messages[-1]}}}) RETURN n")
```

And then I would keep the logic (that was removed in https://github.com/memgraph/memgraph/pull/282/commits/3e99fd0f5ca003200bbc1a14b14d38b5833d9665) to query all the vertices and return them.

The main reason is with the current check the extra messages are allowed, because only the existence of expected messages are checked, but not the absence of extra messages. My original idea was about only to improve the wait time (use `check_one_result_row` only for waiting for the last vertex), and keeping the checking logic as is.

Sorry for the confusion.

File: src/integrations/kafka/consumer.cpp [hunk 23]
```diff
@@ -339,7 +345,22 @@ void Consumer::StartConsuming() {
...
+          spdlog::warn("Saving the commited offset of consumer {} failed: {}", info_.consumer_name,
+                       RdKafka::err2str(err));
+          throw ConsumerCheckFailedException(
+              info_.consumer_name,
+              fmt::format("Couldn't get assignment to commit offsets: '{}'", RdKafka::err2str(err)));
+        }
+
+        for (const auto offset = batch.back().Offset() + 1; auto *partition : partitions) {
+          partition->set_offset(offset);
+        }
```
Comment (antaljanosbenjamin) [inline src/integrations/kafka/consumer.cpp:361]: I think I suggested this method, but right now it struck me that I was wrong. The delivery order of message only guaranteed inside one partition, but not across partitions. So lets assume we have two partitions:
 - partition#1 with messages `1,3,5`
 - partition#2 with messages `2,4,6`
 
 if our batch size is 3, then it is possible that in the first batch we receive only one of the partitions. In case we receive messages `2,4,6`, we will end up totally ignoring messages from partitions#1.
 
 Thinking about it, I think the `assignment` should return the correct offsets, but we have to verify this first. If that's true, then we don't need to set the offset here explicitly.
 
 @antonio2368 @jbajic @kostasrim do you remember why do we decided to set the offset here?

File: src/integrations/kafka/consumer.hpp [hunk 24]
```diff
@@ -146,12 +156,50 @@ class Consumer final : public RdKafka::EventCb {
...
+      }
+    }
+
+    void set_offset(int64_t offset) { offset_ = offset; }
+
+   private:
+    std::optional<int64_t> offset_;
+    std::string consumer_name_;
+  };
+
```
Comment (antaljanosbenjamin) [inline src/integrations/kafka/consumer.hpp:195]: Consider putting the implementation to the source file.

File: src/query/streams.cpp [hunk 25]
```diff
@@ -185,7 +186,27 @@ Streams::Streams(InterpreterContext *interpreter_context, std::string bootstrap_
...
+    }
+  };
+
+  const std::string proc_name = "kafka_set_stream_offset";
```
Comment (antaljanosbenjamin) [inline src/query/streams.cpp:203]:
```suggestion
  constexpr std::string_view proc_name = "kafka_set_stream_offset";
```

File: src/query/streams.cpp [hunk 26]
```diff
@@ -185,7 +186,27 @@ Streams::Streams(InterpreterContext *interpreter_context, std::string bootstrap_
...
+  mgp_proc proc("set_stream_offset", set_stream_offset, utils::NewDeleteResource(), false);
```
Comment (antaljanosbenjamin) [inline src/query/streams.cpp:204]:
```suggestion
  mgp_proc proc(proc_name, set_stream_offset, utils::NewDeleteResource(), false);
```

File: src/integrations/kafka/consumer.cpp (hunk 23 above)
Comment (antonio2368) [reply src/integrations/kafka/consumer.cpp:361]: Not sure, my theory was that we need to manually set it after RebalanceCb sets the offset because the internal structure goes out of sync. That means we cannot use `commitSync` without arguments but I'm not sure if that affects the internal partitions and their assignment. :thinking: It would be best if @kostasrim printed out the values of partition offsets without setting them and seeing if it works correctly.

Comment (kostasrim) [reply src/integrations/kafka/consumer.cpp:361]: Agree. I will check both and update on this. Good catch Benjo!

File: src/integrations/kafka/consumer.cpp [hunk 27]
```diff
@@ -109,8 +109,10 @@ int64_t Message::Timestamp() const {
   return rd_kafka_message_timestamp(c_message, nullptr);
 }
 
+int64_t Message::Offset() const { return message_->offset(); }
```
Comment (antonio2368) [inline src/integrations/kafka/consumer.cpp:112]: maybe add a procedure function to `mgp_message`?

**Review by antonio2368 — APPROVED** [review]

File: src/query/procedure/mg_procedure_impl.cpp [hunk 28]
```diff
@@ -1175,6 +1175,10 @@ mgp_error mgp_date_timestamp(mgp_date *date, int64_t *timestamp) {
   return WrapExceptions([date] { return date->date.MicrosecondsSinceEpoch(); }, timestamp);
 }
 
+mgp_error mgp_message_offset(struct mgp_message *message, int64_t *result) {
+  return WrapExceptions([message] { return message->msg->Offset(); }, result);
+}
```
Comment (antonio2368) [inline src/query/procedure/mg_procedure_impl.cpp:1180]: What about python.

File: src/query/streams.cpp [hunk 29]
```diff
@@ -185,7 +186,27 @@ Streams::Streams(InterpreterContext *interpreter_context, std::string bootstrap_
...
+  auto set_stream_offset = [ictx = interpreter_context, proc_name](mgp_list *args, mgp_graph * /*graph*/,
+                                                                   mgp_result *result, mgp_memory * /*memory*/) {
+    auto *arg_stream_name = procedure::Call<mgp_value *>(mgp_list_at, args, 0);
+    const auto *stream_name = procedure::Call<const char *>(mgp_value_get_string, arg_stream_name);
+    auto *arg_offset = procedure::Call<mgp_value *>(mgp_list_at, args, 1);
+    const auto offset = procedure::Call<int64_t>(mgp_value_get_int, arg_offset);
+    const auto error = ictx->streams.SetStreamOffset(stream_name, offset);
+    if (error.HasError()) {
+      MG_ASSERT(mgp_result_set_error_msg(result, error.GetError().c_str()) == MGP_ERROR_NO_ERROR,
+                "{} unable to set stream offset", proc_name);
```
Comment (antaljanosbenjamin) [inline src/query/streams.cpp:200]:
```suggestion
                "Unable to set procedure error message");
```
In my opinion this assert fails when we cannot set the error message, so the current message might be misleading for the users: it suggest that memgraph crashed because we were unable to set stream offset.

File: src/integrations/kafka/consumer.cpp [hunk 30]
```diff
@@ -358,4 +380,47 @@ void Consumer::StopConsuming() {
   if (thread_.joinable()) thread_.join();
 }
 
+utils::BasicResult<std::string> Consumer::SetConsumerOffsets(int64_t offset) {
+  if (offset == -1) {
```
Comment (antaljanosbenjamin) [inline src/integrations/kafka/consumer.cpp:384]:
```suggestion
  if (is_running_) {
    throw ConsumerRunningException(info_.consumer_name);
  }
  if (offset == -1) {
```

File: src/integrations/kafka/consumer.cpp [hunk 31]
```diff
@@ -339,7 +348,20 @@ void Consumer::StartConsuming() {
 
       try {
         consumer_function_(batch);
-        if (const auto err = consumer_->commitSync(); err != RdKafka::ERR_NO_ERROR) {
+        std::vector<RdKafka::TopicPartition *> partitions;
+        utils::OnScopeExit clear_partitions([&]() { RdKafka::TopicPartition::destroy(partitions); });
+
+        if (const auto err = consumer_->assignment(partitions); err != RdKafka::ERR_NO_ERROR) {
+          spdlog::warn("Saving the commited offset of consumer {} failed: {}", info_.consumer_name,
+                       RdKafka::err2str(err));
```
Comment (antaljanosbenjamin) [inline src/integrations/kafka/consumer.cpp:356]:
```suggestion
```
The exception will be logged in the `catch` part of the `try-catch` block, therefore this log doesn't give any additional value.

**Review by antaljanosbenjamin — COMMENTED** [review]
Thank you Kostas for figuring this out! Only a few changes, all of them are small changes.

**Review by antaljanosbenjamin — APPROVED** [review]
